packages/components/src/input-control/state.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * External dependencies
+ */
+import memize from 'memize';
+/**
+ * WordPress dependencies
+ */
+import { useReducer } from '@wordpress/element';
+
+const initialStateReducer = ( state ) => state;
+
+const initialInputControlState = {
+	_event: {},
+	error: null,
+	initialValue: '',
+	isDirty: false,
+	isDragEnabled: false,
+	isDragging: false,
+	isPressEnterToSubmit: false,
+	value: '',
+};
+
+const actionTypes = {
+	CHANGE: 'CHANGE',
+	DRAG_END: 'DRAG_END',
+	DRAG_START: 'DRAG_START',
+	DRAG: 'DRAG',
+	INVALIDATE: 'INVALIDATE',
+	PRESS_DOWN: 'PRESS_DOWN',
+	PRESS_ENTER: 'PRESS_ENTER',
+	PRESS_UP: 'PRESS_UP',
+	RESET: 'RESET',
+	SUBMIT: 'SUBMIT',
+	UPDATE: 'UPDATE',
+};
+
+export const inputControlActionTypes = actionTypes;
+
+/**
+ * Prepares initialState for the reducer.
+ *
+ * @param {Object} initialState The initial state.
+ * @return {Object} Prepared initialState for the reducer
+ */
+function mergeInitialState( initialState = initialInputControlState ) {
+	const { value } = initialState;
+
+	return {
+		...initialInputControlState,
+		...initialState,
+		initialValue: value,
+	};
+}
+
+/**
+ * Combines multiple stateReducers into a single stateReducer, building
+ * the pipeline to control the flow for state and actions.
+ *
+ * @param  {...any} fns State reducers.
+ * @return {Function} The single combined stateReducer.
+ */
+export const combineStateReducers = ( ...fns ) => {
+	const reducers = fns.flat( Infinity );
+
+	return ( ...args ) => {
+		return reducers.reduceRight( ( state, fn ) => {
+			return { ...state, ...fn( ...args ) };
+		}, {} );
+	};
+};
+
+const memoizedCombineStateReducers = memize( combineStateReducers );
```

Comment:
Per my comment below on combining state reducers at each step of UnitControl > NumberControl > InputControl, there's a chance that this function here become no longer needed, as each component will hold their combined reducer with `useMemo` or equivalent.